src/System.Net.Http/src/System/Net/Http/HttpClientHandler.Core.cs

```diff
@@ -62,5 +64,69 @@ private void CheckDisposedOrStarted()
                 throw new InvalidOperationException(SR.net_http_operation_started);
             }
         }
+
+        private static Exception ValidateAndNormalizeRequest(HttpRequestMessage request)
+        {
+            bool shouldBufferContent = false;
+
+            // Add headers to define content transfer, if not present
+            if (request.HasHeaders && request.Headers.TransferEncodingChunked.GetValueOrDefault())
+            {
+                if (request.Content == null)
+                {
+                    return new HttpRequestException(SR.net_http_client_execution_error,
+                        new InvalidOperationException(SR.net_http_chunked_not_allowed_with_empty_content));
+                }
+
+                // Since the user explicitly set TransferEncodingChunked to true, we need to remove
+                // the Content-Length header if present, as sending both is invalid.
+                request.Content.Headers.ContentLength = null;
+            }
+            else if (request.Content != null && request.Content.Headers.ContentLength == null && request.Headers.TransferEncodingChunked == false)
+            {
+                // We have content, but Content-Length is not set and Transfer-Encoding was explicitly unset.
+                shouldBufferContent = true;
+            }
+            else if (request.Content != null && request.Content.Headers.ContentLength == null && request.Headers.TransferEncodingChunked == null)
+            {
+                // We have content, but Content-Length is not set and Transfer-Encoding was not specified.
+                request.Headers.TransferEncodingChunked = true;
+            }
+
+            if (request.Version.Minor == 0 && request.Version.Major == 1 && request.HasHeaders)
+            {
+                // HTTP 1.0 does not support chunking
+                if (request.Content != null && request.Headers.TransferEncodingChunked == true)
+                {
+                    shouldBufferContent = true;
+                }
+
+                // HTTP 1.0 does not support Expect: 100-continue; just disable it.
+                if (request.Headers.ExpectContinue == true)
+                {
+                    request.Headers.ExpectContinue = false;
+                }
+            }
+
+            if (shouldBufferContent)
+            {
+                long? computedLength = request.Content.GetComputedOrBufferLength();
+
+                if (!computedLength.HasValue)
+                {
+                    request.Content.LoadIntoBufferAsync().GetAwaiter().GetResult();
```

Comment:
We definitely don't want to do this -- we do need to `await`. Here is what I'm hoping for:

- Rename the platform-specific `SendAsync` to `SendWithPlatformHandlerAsync`
- Make a new generic, non-async `HttpClientHandler.SendAsync` in HttpClientHandler.cs. Have this check `TransferEncodingChunked` and:
   - If it needs to buffer things, send to an async helper method `SendWithContentLengthAsync` which buffers, sets `ContentLength`, and finally awaits `SendWithPlatformHandlerAsync`.
    - Otherwise, send directly to `SendWithPlatformHandlerAsync`.

This way we use async when we need the I/O to buffer, but bypass it when not needed.

Reply:
Derp, I understand what you meant now :|

Reply:
Fixed